haskell-src/Concordium/Types/SeedState.hs

```diff
@@ -52,36 +51,103 @@ withIsSeedStateVersionFor :: SProtocolVersion pv -> (IsSeedStateVersion (SeedSta
 withIsSeedStateVersionFor spv = withSingI (sSeedStateVersionFor spv)
 
 -- |State for computing the leadership election nonce.
-data SeedState (ssv :: SeedStateVersion) = SeedState
-    { -- |Number of slots in an epoch. This is derived from genesis
-      -- data and must not change.
-      epochLength :: !(Conditionally (SupportsEpochLength ssv) EpochLength),
-      -- |Current epoch
-      epoch :: !Epoch,
-      -- |Current leadership election nonce
-      currentLeadershipElectionNonce :: !LeadershipElectionNonce,
-      -- |The leadership election nonce updated with the block nonces
-      -- of blocks in the first 2/3 of the current epoch.
-      updatedNonce :: !Hash
-    }
-    deriving (Eq, Show)
+--
+-- (Implementation note: the constructors have equational constraints so that they have the same
+-- return type. This allows them to have common field names (e.g. 'epoch'), which is not
+-- allowed if the return types differ (e.g. @SeedState 'SeedStateVersion0@ and
+-- @SeedState 'SeedStateVersion1@).)
```

Comment:
Does this provide a lot of convenience? 

What happens to `epochTransitionTriggered`? In normal ADTs you would get a partial function from a field that only appears in one of the variants. Which is annoyingly error prone. 

Is that better here in that you get a function
```
epochTransitionTriggered :: (ssv ~ 'SeedStateVersion1) -> SeedStateVersion ssv -> Bool
```
?

Reply:
Yes. Instead of throwing an exception at runtime then it will error out at compile time telling you that the the constraints can't be met (if you were to define that function like :

```
epochTransitionTriggered :: (ssv ~ 'SeedStateVersion1) => SeedStateVersion ssv -> Bool

```

Reply:
Not except for naming.

Reply:
I have reworked to use lenses instead that are type safe.